python/observation-publisher/Makefile

```diff
@@ -4,6 +4,12 @@ ENVIRONMENT_CONFIG = "example-override"
 setup:
 	@echo "Setting up environment..."
 	@pip install -r requirements.txt --use-pep517
+	@pip install -r requirements-dev.txt
+
+.PHONY: pip-compile
+pip-compile:
+	@echo "Compiling requirements..."
+	@python -m piptools compile
```

Comment:
just curious, what is the purpose of this compiling action?

Reply:
Let say we are writing our own requirements.txt, and pin all the dependencies we require on specific versions. Unless we pin the transitive dependencies as well, the dependencies installed may differ every time we build the image / install the dependencies.

We could do a pip freeze and capture all the dependencies, but it would be harder to update the dependencies at a later time. You also won't have good visibility on which dependencies are core dependencies, and which are transitive dependencies.

What piptools does is that, it takes in an input file, where the dependencies version may or may not be fixed, then generate a requirements.txt that has the exact dependency version, along with comments to illustrate the dependency hierarchy.  One advantage of doing so, is that, let say you update a dependency, such as google-cloud-bigquery, you can run pip-compile to know what are the dependencies that could be upgraded / are incompatible.

This approach is certainly not suitable for libraries, but for standalone application, it's a fairly common practice.